metricflow/model/validations/identifiers.py

```diff
@@ -119,3 +123,86 @@ def validate_model(model: UserConfiguredModel) -> List[ValidationIssue]:  # noqa
 
         return issues
+
+
+@dataclass(frozen=True)
+class SubIdentifierContext:
+    """Organizes the context behind identifiers and their sub-identifiers."""
+
+    data_source_name: str
+    identifier_reference: IdentifierReference
+    sub_identifier_names: Tuple[str, ...]
+
+
+class IdentifierConsistencyRule(ModelValidationRule):
+    """Checks identifiers with the same name are defined with the same set of sub-identifiers in all data sources"""
+
+    @staticmethod
+    def _get_sub_identifier_names(identifier: Identifier) -> Sequence[str]:
+        sub_identifier_names = []
+        sub_identifier: CompositeSubIdentifier
+        for sub_identifier in identifier.identifiers or []:
+            if sub_identifier.name:
+                sub_identifier_names.append(sub_identifier.name.element_name)
+            elif sub_identifier.ref:
+                sub_identifier_names.append(sub_identifier.ref)
+        return sub_identifier_names
+
+    @staticmethod
+    def _get_sub_identifier_context(data_source: DataSource) -> Sequence[SubIdentifierContext]:
+        contexts = []
+        for identifier in data_source.identifiers or []:
+            contexts.append(
+                SubIdentifierContext(
+                    data_source_name=data_source.name,
+                    identifier_reference=identifier.name,
+                    sub_identifier_names=tuple(IdentifierConsistencyRule._get_sub_identifier_names(identifier)),
+                )
+            )
+        return contexts
+
+    @staticmethod
+    @validate_safely(whats_being_done="running model validation to ensure identifiers have consistent sub-identifiers")
+    def validate_model(model: UserConfiguredModel) -> List[ValidationIssueType]:  # noqa: D
+        issues: List[ValidationIssueType] = []
+        # The expected sub identifiers that an identifier has.
+        identifier_to_sub_identifiers: OrderedDict[str, Tuple[str, ...]] = OrderedDict()
+        # The identifiers that have different sub-identifiers in different data sources.
+        identifiers_with_inconsistent_sub_identifiers = []
+
+        identifier_contexts = []
+        issues = []
+        for data_source in model.data_sources or []:
+            sub_identifier_contexts = IdentifierConsistencyRule._get_sub_identifier_context(data_source)
+            for sub_identifier_context in sub_identifier_contexts:
+                identifier_contexts.append(sub_identifier_context)
+
+                identifier_reference = sub_identifier_context.identifier_reference
+                if identifier_reference.element_name not in identifier_to_sub_identifiers:
+                    identifier_to_sub_identifiers[
+                        identifier_reference.element_name
+                    ] = sub_identifier_context.sub_identifier_names
+
+                elif (
+                    identifier_to_sub_identifiers[identifier_reference.element_name]
+                    != sub_identifier_context.sub_identifier_names
```

Comment:
So these are order-dependent? Is that necessary? Assuming this is a direct port from the original validation, we should just go with it, as relaxing that restriction seems like a discussion beyond the scope of this PR.

Reply:
This is a direct port from the original validation. We can re-address this later.